### PR TITLE
Syntax à la carte

### PR DESCRIPTION
**Pull request by tclem**
This brings together the [data types à la carte](http://www.cs.ru.nl/~W.Swierstra/Publications/DataTypesALaCarte.pdf) technique with the existing abstract interpretation work allowing à la carte definition of syntax. Some very interesting things fall out from this approach!

Huge 🎩  to @robrix for 🍐 ing with me on much of this!

### CODE REVIEW

File: src/Abstract/Interpreter/Tracing.hs [hunk 1]
```diff
@@ -30,27 +31,51 @@ type TraceResult l t v f = Final (TracingInterpreter l t v f) v
...
+--    evalReach @Precise @(Value Syntax Precise) @Syntax (makeLam "x" (var "x") # true)
+
+evalTrace :: forall l v s
+          . ( Ord v, Ord1 s, Ord1 (Cell l)
+            , MonadAddress l (Eff (TraceInterpreter l (Term s) v))
+            , MonadPrim v (Eff (TraceInterpreter l (Term s) v))
+            , MonadGC l v (Eff (TraceInterpreter l (Term s) v))
+            , Semigroup (Cell l v)
+            , Eval v (Eff (TraceInterpreter l (Term s) v)) s s
+            )
```
Comment (robrix) [inline src/Abstract/Interpreter/Tracing.hs:51]: I really dig how you’ve formatted these contexts 👍 Thank you!

File: stack.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+packages:
+- .
+- location: vendor/effects
+  extra-dep: true
+- vendor/prettyprinter/prettyprinter
+- vendor/prettyprinter/prettyprinter-ansi-terminal
+- vendor/prettyprinter/prettyprinter-compat-annotated-wl-pprint
+- vendor/prettyprinter/prettyprinter-compat-ansi-wl-pprint
+- vendor/prettyprinter/prettyprinter-compat-wl-pprint
+- vendor/prettyprinter/prettyprinter-convert-ansi-wl-pprint
```
Comment (robrix) [inline stack.yaml]: We’re not using `prettyprinter` any more, so I think we can 🔥 these (and the submodule).

File: src/Abstract/Value.hs [hunk 3]
```diff
@@ -1,182 +1,58 @@
...
-  liftPretty2 pL plL pA plA = list . map (liftPretty prettyAddress (list . map prettyAddress)) . Map.toList . unEnvironment
-    where prettyAddress = liftPretty2 pL plL pA plA
+-- instance Show1 Value where
+--   liftShowsPrec spL slL = go
+--     where go d v = case v of
+--             I a -> showsUnaryWith showsPrec "I" d a
+--             Closure s t e -> showsConstructor "Closure" d [flip showsPrec s, flip showsPrec t, flip (liftShowsPrec2 spL slL go (showListWith (go 0))) e]
+--
+-- instance Show l => Show (Value l) where
+--   showsPrec = showsPrec1
```
Comment (robrix) [inline src/Abstract/Value.hs]: 🔥, please 🙏

File: src/Abstract/Value.hs [hunk 4]
```diff
@@ -1,182 +1,58 @@
...
-instance (MonadAddress l m, MonadStore l (Value l) m, MonadEnv l (Value l) m, MonadFail m, Semigroup (Cell l (Value l))) => MonadValue l (Value l) (Term Prim) m where
-  rec ev name e0 =  do
-    a <- alloc name
-    v <- localEnv (envInsert name (a :: Address l (Value l))) (ev e0)
-    assign a v
-    return v
+-- instance Eq2 Value where
+  -- liftEq2 eql1 eql2 = go
+  --   where go v1 v2 = case (v1, v2) of
+  --           _ -> False
```
Comment (robrix) [inline src/Abstract/Value.hs]: 🔥, please 🙏

File: src/Abstract/Syntax.hs [hunk 5]
```diff
@@ -1,283 +1,437 @@
...
+--
+-- instance Num a => Num (Term a) where
+--   fromInteger = prim . fromInteger
+--
+--   -- signum = In . Op1 Signum
+--   -- abs    = In . Op1 Abs
+--   -- negate = In . Op1 Negate
+--   -- (+) = (In .) . Op2 Plus
+--   -- (-) = (In .) . Op2 Minus
+--   -- (*) = (In .) . Op2 Times
```
Comment (robrix) [inline src/Abstract/Syntax.hs]: 🔥, please 🙏

File: src/Abstract/Syntax.hs [hunk 6]
```diff
@@ -1,283 +1,437 @@
...
+    assign a (TVar tvar)
+    outTy <- localEnv (envInsert name (a :: Address Monovariant Type)) (ev body)
+    return (TVar tvar :-> outTy)
+
+instance FreeVariables1 Lambda where
+  liftFreeVariables f (Lambda name body) = delete name (f body)
+
+instance (Monad m, Eval v m s Lambda) => EvalCollect l v m s Lambda
+
+-- Recursive lambdas
```
Comment (robrix) [inline src/Abstract/Syntax.hs]: `Rec` is a binder, but not actually a lambda—it’s more like `letrec` in ML (or just `let` in Haskell), i.e. it introduces a binding which is in scope in its definition. Defining a recursive function requires both a `Rec` and a `Lambda`; if you only have a `Rec`, you can define recursively-constructed values, but not functions.

File: src/Abstract/Interpreter/Dead.hs [hunk 7]
```diff
@@ -39,23 +41,28 @@ subterms term = para (foldMap (uncurry ((<>) . point))) term <> point term
 
 
 -- Dead code analysis
-
-evalDead :: forall l v. (MonadAddress l (Eff (DeadCodeInterpreter l (Term Prim) v)), MonadValue l v (Term Prim) (Eff (DeadCodeInterpreter l (Term Prim) v)), MonadPrim v (Eff (DeadCodeInterpreter l (Term Prim) v)), Semigroup (Cell l v)) => Eval (Term Prim) (DeadCodeResult l (Term Prim) v)
-evalDead e0 = run @(DeadCodeInterpreter l (Term Prim) v) $ do
+-- Example:
+--    evalDead @Precise @(Value Syntax Precise) @Syntax (if' true (Abstract.Syntax.int 1) (Abstract.Syntax.int 2))
```
Comment (robrix) [inline src/Abstract/Interpreter/Dead.hs:45]: 💖

**Review by robrix — APPROVED** [review]
👍 this is great, @tclem! Just a few notes.

File: abstract-interpretation.cabal [hunk 8]
```diff
@@ -14,26 +14,30 @@ cabal-version:       >=1.10
...
                      , Abstract.Primitive
                      , Abstract.Set
                      , Abstract.Store
                      , Abstract.Syntax
+                     , Abstract.Term
                      , Abstract.Type
-                     , Abstract.Util
                      , Abstract.Value
                      , Control.Effect
+                     , Data.Functor.Classes.Show.Generic
```
Comment (robrix) [inline abstract-interpretation.cabal]: @tclem: It just occurred to me that this is gonna collide with any other package providing a module by the same name.

I think our best bet for the time being is to put this into an `other-modules` section for the library, which will prevent collisions.

Comment (tclem) [reply abstract-interpretation.cabal]: 👍  Good call. I brought over the generic `Eq(1)` and `Ord(1)` work you've done as well and put them all in the `other-modules` section.

**Pull request merged**